### PR TITLE
Introduce a new URL for library pages, at index.html

### PR DESCRIPTION
**Pull request by srawlins**
Fixes https://github.com/dart-lang/dartdoc/issues/1346

This "moves" the URL of each library from, e.g.
'package-two_lib2/package-two_lib2-library.html' to 'package-two_lib2/index.html', such that we can have links with 'package-two_lib2'. The old URLs are preserved via redirecting HTML files.

(HTTP redirects are more standard and recommended, but dartdoc does not contain an HTTP server; HTTP redirects could be implemented at the hosting server, which is different for different doc hosts (Google Storage, Firebase, ...). Instead we use simpler HTML redirects. In fact, a given backend host could use the HTML redirecting files as a database to inform HTTP redirect rules, that need to be implemented differently for each backend. So dartdoc now produces redirecting files, but a given backend could instead return an HTTP 301 Permanent Redirect response when requesting one of the redirecting files.)

This adds a negligible number of files (1 per documented library), and a negligible number of bytes (tiny redirect files), but actually is probably a net negative in terms of bytes, as many links are now shortened, removing the HTML file name of library URLs.

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/blob/main/docs/External-Package-Maintenance.md#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

**Comment by gspencergoog** [comment]
Would it make sense to have a mode that doesn't emit redirects at all, and renames the output files for all the symbol types that aren't suffixed, like variable, enum, and function (so the top level "index" variable doesn't collide with "index.html", etc.)? Then that mode could be used for "new" output, and the redirects could be used for "legacy" output.  I'd hate to think that we're stuck emitting ever larger numbers of redirect files every time we rename something in the output tree.

**Comment by srawlins** [comment]
We could have a mode that stops emitting the redirecting files. You'd enable it whenever you're comfortable with some % of broken links coming from StackOverflow, articles, etc.

**Comment by gspencergoog** [comment]
> We could have a mode that stops emitting the redirecting files. You'd enable it whenever you're comfortable with some % of broken links coming from StackOverflow, articles, etc.

I was imagining using it mainly for new packages, so in those cases there wouldn't be any broken links.

**Comment by srawlins** [comment]
Ah, we could do that. It wouldn't be used very much, as a new package author would have to know about it and include it in their dartdoc_options.yaml, or it could be the default and offer the "please don't break old links" mode be opt-in.

**Comment by gspencergoog** [comment]
Okay, maybe not worth the effort, then.  Having it be the default might cause havoc (since people wouldn't know to turn it on and make lots of broken links), and having it not be the default means it won't be used much.

File: lib/src/generator/templates.aot_renderers_for_html.dart [hunk 1]
```diff
@@ -1137,6 +1137,29 @@ String renderLibrary(LibraryTemplateData context0) {
   return buffer.toString();
 }
 
+String renderLibraryRedirect(LibraryTemplateData context0) {
+  final buffer = StringBuffer();
+  buffer.write('''<!DOCTYPE html>
+<html lang="en">
```
Comment (kevmoo) [inline lib/src/generator/templates.aot_renderers_for_html.dart:1143]: Might could add a `canonical` tag, too? For the search engines?

Comment (kevmoo) [reply lib/src/generator/templates.aot_renderers_for_html.dart:1143]: https://developers.google.com/search/docs/crawling-indexing/consolidate-duplicate-urls

Comment (srawlins) [reply lib/src/generator/templates.aot_renderers_for_html.dart:1143]: So good! Thanks Kevin, I love this. Done.

**Comment by srawlins** [comment]
Ready for review @dart-lang/analyzer-team . Big feature for dartdoc.

**Review by bwilkerson — APPROVED** [review]

**Pull request merged**